File: Classes/Joint.m
classdef Joint < JointInterface
    % This is a joint, which is represented by a single standard Denavit
    % Hartenberg frame
    
    %% Private Properties
    properties (Access = private) 
        a double; %DH Parameter (in m)
        d double; %DH Parameter (in m)
        theta double; %DH Parameter (in radians)
        alpha double; %DH Parameter (in radians)
        alpha2 double; %DH Parameter (in radians) -- NOT USED IN THIS CHILD< BUT NECESSARY FOR INHERITENCE
        Frame double = zeros(4,4); % DH Frame
    end
    
    %% Public Methods
    methods
        
        %Constructor
        function obj = Joint(theta,alpha,a,d)
            %JOINT Construct an instance of Joint

            % Inject DH parameters
            obj.a = a;
            obj.d = d;
            obj.theta = deg2rad(theta);
            obj.alpha = deg2rad(alpha);

            % Evaluate Frame
            obj = evaluateFrame(obj);
        end

        function obj = UpdateAngles(obj,Angles)
            % Update the angles and re-evaluate the frame

            obj.theta = deg2rad(Angles(1));
            obj.alpha = deg2rad(Angles(2));

            % Evaluate Frame
            obj = evaluateFrame(obj);

        end

        %% Accessors
        function Frame = getFrame(obj)
            
            % Accessor to retrieve the private property DH frame
            Frame = obj.Frame;
        end

    end    
    
    
    %% Private Methods
    methods (Access = private)
        function obj = evaluateFrame(obj)
            %evaluateFrame uses the internal properties to evaluate the DH
            %frame associated with this Joint.

            obj.Frame = [cos(obj.theta),-sin(obj.theta)*cos(obj.alpha),sin(obj.theta)*sin(obj.alpha),obj.a*cos(obj.theta);...
            sin(obj.theta),cos(obj.theta)*cos(obj.alpha),-cos(obj.theta)*sin(obj.alpha),obj.a*sin(obj.theta);...
            0,sin(obj.alpha),cos(obj.alpha),obj.d;...
            0,0,0,1];

            %set any tiny elements to 0
            obj.Frame(abs(obj.Frame) < (1.0e-10)) = 0;
        end
    
    end

%% End Class definition
end

File: cpp/MagnetSolver/Joint.cpp
// This is a joint, which is represented by a single standard Denavit Hartenberg frame

#define _USE_MATH_DEFINES																									// To ensure that M_PI works
#include "Joint.h"
#include <cmath>																											// For M_PI


// Constructor
Joint::Joint(float theta, float alpha, float a, float d)
	: _theta(deg2rad(theta)), _alpha(deg2rad(alpha)), _a(a), _d(d), _Frame(Eigen::Matrix4f::Zero()) {						// Member initialisation list - inits the class members. Note angles are immediately converted to rad
	
	Joint::evaluateFrame();
}

// Defualt constructor																									
Joint::Joint()																												// COnstructor overloading - this object is instantiated when no iput arguements are provided
	: _theta(0.0f), _alpha(0.0f), _a(0.0f), _d(0.0f), _Frame(Eigen::Matrix4f::Identity()) {
	evaluateFrame();
}

// Destructor
Joint::~Joint() {
	// no action necessary.
};


// Return a reference to the DH Frame
Eigen::Matrix4f& Joint::getFrame() {

	return _Frame;
};


// Evaluate the DH Frame
void Joint::evaluateFrame() {

	// overwrite frame with newly calculated parameters.
	_Frame << cos(_theta), -sin(_theta) * cos(_alpha), sin(_theta)* sin(_alpha), _a* cos(_theta),
		sin(_theta), cos(_theta)* cos(_alpha), -cos(_theta) * sin(_alpha), _a* sin(_theta),
		0, sin(_alpha), cos(_alpha), _d,
		0, 0, 0, 1;

	// Set any tiny elements to zero
	_Frame = _Frame.unaryExpr([](float elem) { return fabs(elem) < 1.0e-7 ? 0.0f : elem; });

	// Eigen::Matrix.unaryExpr applies an operation to each element in the array.
	// The operation is a lambda
	// The lambda checks whether the absoulte value is less than small number, and returns 0.0 if so. if not, it returns the element
}


// Update the angles
void Joint::UpdateAngles(std::vector<float> Angles) {

	_theta = deg2rad(Angles[0]);
	_alpha = deg2rad(Angles[1]);

	Joint::evaluateFrame();																									//revaluate frame after joint angle update.
}


// Convert degrees to radians
float Joint::deg2rad(float degrees) {
	return degrees * M_PI / 180.0f;
}